extensions/nanoarrow_ipc/src/nanoarrow/nanoarrow_ipc.h

```diff
@@ -219,6 +227,60 @@ ArrowErrorCode ArrowIpcDecoderDecodeArray(struct ArrowIpcDecoder* decoder,
                                           struct ArrowArray* out,
                                           struct ArrowError* error);
 
+/// \brief An user-extensible input data source
+struct ArrowIpcInputStream {
+  /// \brief Read up to buf_size_bytes from stream into buf
+  ///
+  /// The actual number of bytes read is placed in the value pointed to by
+  /// size_read_out. Returns NANOARROW_OK on success.
+  ArrowErrorCode (*read)(struct ArrowIpcInputStream* stream, void* buf,
```

Comment:
I suppose:

- Perhaps consider a concrete type (char*, uint8_t*) for buf so it's harder to accidentally pass an arbitrary pointer to it?
- What do you think about `char** buf` (i.e., letting the stream do the allocation)? That would let you do things like read an in-memory buffer by slicing instead of copying (though I suppose lifetimes may get confusing there).